### PR TITLE
Exit signal gets its own trait

### LINKED ISSUE #424: Panic while CTRL-C during sync
Version: f77b3e32
Run Polkadot, let it sync to the point it can import new blocks.
Can reproduce: not 100% of the time but often.

<details>
<summary>Trace</summary>

```
2018-07-26 10:45:05 Imported #826942 (8891…e277)
^C2018-07-26 10:45:08 Idle (6 peers), best: #826942 (8891…e277)
2018-07-26 10:45:12 Imported #826943 (78f6…bab9)

====================

stack backtrace:
   0:        0x109782434 - backtrace::backtrace::trace::h9908331efcb29cea
                        at /Users/will/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/backtrace/mod.rs:42
   1:        0x10977d3bc - backtrace::capture::Backtrace::new_unresolved::h743fddbdb5d489cb
                        at /Users/will/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/capture.rs:88
   2:        0x10977d31e - backtrace::capture::Backtrace::new::hc75417b04e0c260d
                        at /Users/will/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/capture.rs:63
   3:        0x1078675ec - substrate_cli::panic_hook::panic_hook::hea28e132763072f4
                        at substrate/cli/src/panic_hook.rs:51
   4:        0x10785e428 - core::ops::function::Fn::call::h9cbf7a49a5f3b9a5
                        at /Users/travis/build/rust-lang/rust/src/libcore/ops/function.rs:73
   5:        0x109bbf0a8 - std::panicking::rust_panic_with_hook::h67cebecd26baa567
                        at libstd/panicking.rs:479
   6:        0x109bbebfc - std::panicking::continue_panic_fmt::h94e33f67d6da6141
                        at libstd/panicking.rs:390
   7:        0x109bbeae8 - rust_begin_unwind
                        at libstd/panicking.rs:325
   8:        0x109c02c71 - core::panicking::panic_fmt::ha2c495619d2ab627
                        at libcore/panicking.rs:77
   9:        0x10945c94b - core::result::unwrap_failed::h85c28f2ae30f3b4e
                        at /Users/travis/build/rust-lang/rust/src/libcore/macros.rs:26
  10:        0x106f60565 - _$LT$core..result..Result$LT$T$C$$u20$E$GT$$GT$::expect::h8138ca00509a440b
                        at /Users/travis/build/rust-lang/rust/src/libcore/result.rs:809
  11:        0x106f721e5 - _$LT$polkadot..Worker$u20$as$u20$polkadot_cli..Worker$GT$::exit_only::_$u7b$$u7b$closure$u7d$$u7d$::h67fad022e6d682d6
                        at polkadot/src/main.rs:52
  12:        0x106fc2a03 - ctrlc::CtrlC::set_handler_with_polling_rate::_$u7b$$u7b$closure$u7d$$u7d$::hff0781889c0fcc06
                        at /Users/will/.cargo/git/checkouts/rust-ctrlc-b59d90608a6f68aa/b523017/src/lib.rs:119
  13:        0x106fd8b6f - std::sys_common::backtrace::__rust_begin_short_backtrace::h84477289f97de49f
                        at /Users/travis/build/rust-lang/rust/src/libstd/sys_common/backtrace.rs:136
  14:        0x106f69b6f - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h2918353078332d66
                        at /Users/travis/build/rust-lang/rust/src/libstd/thread/mod.rs:409
  15:        0x106fc2dff - _$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h4e9829448da4212a
                        at /Users/travis/build/rust-lang/rust/src/libstd/panic.rs:313
  16:        0x106f9720c - std::panicking::try::do_call::h09ace90ad68b7e23
                        at /Users/travis/build/rust-lang/rust/src/libstd/panicking.rs:310
  17:        0x109bcb13e - __rust_maybe_catch_panic
                        at libpanic_unwind/lib.rs:106
  18:        0x106f97114 - std::panicking::try::h1d89c070fa60bde5
                        at /Users/travis/build/rust-lang/rust/src/libstd/panicking.rs:289
  19:        0x106fc2e3f - std::panic::catch_unwind::h22be69cb0f759834
                        at /Users/travis/build/rust-lang/rust/src/libstd/panic.rs:392
  20:        0x106f699af - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::ha73ac04ad26cc453
                        at /Users/travis/build/rust-lang/rust/src/libstd/thread/mod.rs:408
  21:        0x106f69cc7 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::ha776c5bb3495b951
                        at /Users/travis/build/rust-lang/rust/src/liballoc/boxed.rs:640
  22:        0x109ba53bb - std::sys::unix::thread::Thread::new::thread_start::he77c77667b31978e
                        at libstd/sys_common/thread.rs:24
  23:     0x7fff67811660 - _pthread_body
  24:     0x7fff6781150c - _pthread_start

Thread '<unnamed>' panicked at 'Error sending exit notification: ()', libcore/result.rs:945

This is a bug. Please report it at:

    https://github.com/paritytech/polkadot/issues/new
```

</details>

### PR DESCRIPTION
**Pull request by arkpar**
Fixes #424

### CODE REVIEW

File: substrate/cli/src/lib.rs [hunk 1]
```diff
@@ -85,11 +85,19 @@ pub struct VersionInfo {
...
 	/// Service mode requested. Caller should start the service.
-	RunService(FactoryFullConfiguration<F>),
+	RunService((FactoryFullConfiguration<F>, E)),
+}
+
+/// Something that can be converted into an exit signal.
+pub trait IntoExit {
+	/// Exit signal type.
+	type Exit: Future<Item=(),Error=()> + Send + 'static;
+	/// Convet into exit signal.
```
Comment (pepyakin) [inline substrate/cli/src/lib.rs]: typo: `Convert`

File: polkadot/src/main.rs [hunk 2]
```diff
@@ -38,11 +38,9 @@ mod vergen {
 
 // the regular polkadot worker simply does nothing until ctrl-c
 struct Worker;
-impl cli::Worker for Worker {
-	type Work = Self::Exit;
+impl cli::IntoExit for Worker {
 	type Exit = future::MapErr<oneshot::Receiver<()>, fn(oneshot::Canceled) -> ()>;
-
-	fn exit_only(&self) -> Self::Exit {
+	fn into_exit(self) -> Self::Exit {
```
Comment (pepyakin) [inline polkadot/src/main.rs]: nice!

File: polkadot/cli/src/lib.rs [hunk 3]
```diff
@@ -85,9 +79,9 @@ pub fn run<I, T, W>(args: I, worker: W, version: cli::VersionInfo) -> error::Res
 	W: Worker,
 {
 
-	match cli::prepare_execution::<service::Factory, _, _, _, _>(args, worker.exit_only(), version, load_spec, "parity-polkadot")? {
+	match cli::prepare_execution::<service::Factory, _, _, _, _>(args, worker, version, load_spec, "parity-polkadot")? {
 		cli::Action::ExecutedInternally => (),
-		cli::Action::RunService(mut config) => {
+		cli::Action::RunService((mut config, worker)) => {
```
Comment (pepyakin) [inline polkadot/cli/src/lib.rs]: Hm, I'm not sure how it works. The second item in tuple of `Action::RunService` is bound to be `IntoExit`, why it's called `worker` here? Why `worker.configuration()` works here?

Comment (arkpar) [reply polkadot/cli/src/lib.rs]: `Action` returns back the same thing that was passed to `prepare_execution` if it was not used.

File: polkadot/collator/src/lib.rs [hunk 4]
```diff
@@ -211,12 +211,21 @@ struct CollationNode<P, E> {
 	key: Arc<ed25519::Pair>,
 }
 
+impl<P, E> IntoExit for CollationNode<P, E> where
+	P: ParachainContext + Send + 'static,
+	E: Future<Item=(),Error=()> + Send + Clone + 'static
```
Comment (rphmeier) [inline polkadot/collator/src/lib.rs]: `Clone` bound no longer required

**Pull request merged**